util/journaldb/src/overlayrecentdb.rs

```diff
@@ -279,11 +280,19 @@ impl JournalDB for OverlayRecentDB {
 	fn earliest_era(&self) -> Option<u64> { self.journal_overlay.read().earliest_era }
 
 	fn state(&self, key: &H256) -> Option<Bytes> {
-		let journal_overlay = self.journal_overlay.read();
 		let key = to_short_key(key);
-		journal_overlay.backing_overlay.get(&key, EMPTY_PREFIX)
-			.or_else(|| journal_overlay.pending_overlay.get(&key).map(|d| d.clone()))
-			.or_else(|| self.backing.get_by_prefix(self.column, &key[0..DB_PREFIX_LEN]).map(|b| b.to_vec()))
+		let maybe_state_data = {
+			let journal_overlay = self.journal_overlay.try_read_for(Duration::from_secs(2))?;
+			journal_overlay.backing_overlay.get(&key, EMPTY_PREFIX)
+				.or_else(|| journal_overlay.pending_overlay.get(&key).map(|d| d.clone()))
+		};
+		match maybe_state_data {
```

Comment:
The last match/this could be removed with two chained `or_else` as it was before as I prefer.

Reply:
The idea is to release the read lock as quickly as possible; the lock [should be fair](https://docs.rs/parking_lot/0.10.0/parking_lot/type.RwLock.html#fairness) but we've seen it lock up other threads, hence this change. One conjecture is that the multiple levels of locking we have going on in this code path messes with the fairness, but tbh I do not know.
I do know that this change alone (i.e. without any of the other changes in this PR or @ordian's fix) made a difference in responsiveness.

Reply:
ok, just replace the last match with `or_else` and I'm fine with it, the current code to is fine too I'm just a bit picky on `Some => Some` :)